Wp 777 two pzp name handling

habibvirji opened the pull request:
PZP name generated is md5 hash of devicename, osname, totalmem, osarch, ostype, corrent working directory, webinos type and array of network interface. These are static information and should help PZP to identity itself.
PZP with same name in PZ are assigned new id by PZH. Storage wise it is going to use above id but while communicating with PZH it is going to use PZH assigned id.

Jira Issue: WP-777

habibvirji committed:
Two PZP name handling

PZP name generated is md5 hash of devicename, osname, totalmem, osarch, ostype, corrent working directory, webinos type and array of network interface. These are static information and should help PZP to identity itself.
PZP with same name in PZ are assigned new id by PZH. Storage wise it is going to use above id but while communicating with PZH it is going to use PZH assigned id.

Jira Issue: WP-777

habibvirji commented:
@TobyEalden @jplyle I do not have VM to get device name same on both PZPs. Can you check this at your convenience if possible.

jplyle commented:
@habibvirji I'm testing it now.

jplyle commented:
Hm.  I'm not totally sure this works.  Firstly, my two VMs generate the same ID because they are identical VMs.  That's OK, although possibly the 'device ID' should just be a randomly generated integer value rather than a hash of anything.

When I connect both PZPs, it the PZH seems less confused, although service discovery appears to fail in the PZP testbed.  Also, I get the old errer:

/home/johl/git-repos/Webinos-Platform/webinos/core/api/servicedisco/lib/rpc_servicedisco.js:270
        var count = this.remoteServiceObjects[address].length;
                                                      ^
TypeError: Cannot read property 'length' of undefined
    at Discovery.removeRemoteServiceObjects (/home/johl/git-repos/Webinos-Platform/webinos/core/api/servicedisco/lib/rpc_servicedisco.js:270:49)
    at Pzh.removeRoute (/home/johl/git-repos/Webinos-Platform/webinos/core/pzh/lib/pzh_tlsSessionHandling.js:252:41)
    at CleartextStream.Provider.fetchPzh (/home/johl/git-repos/Webinos-Platform/webinos/core/pzh/lib/pzh_provider.js:202:39)
    at CleartextStream.EventEmitter.emit (events.js:93:17)
    at SecurePair.destroy (tls.js:956:22)
    at process.startup.processNextTick.process._tickCallback (node.js:244:9)

I can't confirm that this works.  Could you rebase?

habibvirji commented:
Okay will rebase.

PZP cannot be random generate id as PZP needs to generate same id to load configuration when restarted. Otherwise it will have to recreate configuration if it cannot generate old id.

jplyle commented:
Habib - I'm confused.  Can't it generate a random ID and then reuse this?  Not that it really matters that much...

habibvirji commented:
How will it relocate back pzp folder if it generates random id?

habibvirji commented:
PZP details are in that folder, so the id generated has to be read then from elsewhere. This id needs to be generated everytime to find out PZP folder. Not just first time when PZP is started.

habibvirji committed:
Pzh Assigned Id in Session Handling

Jira Issue:WP-777

jplyle commented:
Oh, I see.  

Is there a reason why the PZP ID is used as a directory name?  This is probably far too much work to be worth changing, but it seems like you don't need subfolders from ~/.webinos/ anymore.  The PZP could look in ~/.webinos/ and load everything out of that path, and well-known subfolders.  The directory isn't shared by other PZPs or PZHs, after all.  Or at least it shouldn't be...

habibvirji commented:
Agreed not relevant for the PZP although the origin was due to starting 2 PZP on the same machine via forcedDeviceName option. 

The code is same for the PZH (as provider will have multiple PZH running). Will require separating codes for both the cases.

habibvirji committed:
Webinos PZP path change

PZP friendly name change if same friendly name is used.
PZP default folder changed to .webinos.
Files stored with webinosId is moved to different file names.
Webinos Device Name is MD5 of cwd and device name

Jira Issue: WP-777

habibvirji commented:
@jplyle @TobyEalden can you please check these functionality. Changes .webinos folder and if 2 pzp with same friendlyname if present changes them. (I do not have 2 machines to test it).

jplyle commented:
I'm getting the same problem with this as I'm getting with your other PRs - enrolment fails on Chrome now.

habibvirji commented:
Have you run grunt for each PR? webinos.js differs for all PRs.

jplyle commented:
These are the exact steps I'm running (the remote 'habib' is set to your repository) 

git checkout -- webinos_config.json package.json
git checkout master
git checkout -- webinos_config.json package.json
git fetch habib
git checkout habib/WP-777-two-pzp-name-handling
rm -rf ~/git-repos/Webinos-Platform/node_modules
git checkout -- node_modules/expat2.js node_modules/xmldigsig.js
npm install

...

Sorry, I was having an issue with timezones.

jplyle commented:
Please: do not merge until WP-718 is merged.

jplyle commented:
This works a charm.  When I try to enrol two identically-named virtual machine based PZPs, one appears as "Linux Device" and the other appears as "Linux Device#45".  If that's the intention, it works for me.

habibvirji commented:
Perfect. I am using random value if there are devices with the same friendly name. As I said previously this is just to avoid confusion. User can change friendly name anytime via setfriendlyname interface. If it is okay can you please merge this.
